server/transcription_api.go

```diff
@@ -0,0 +1,192 @@
+		clientState.EndAt = time.Now().UnixMilli()
+
+		if state.Call.Recording != nil && state.Call.Recording.EndAt == 0 {
+			if _, _, err := p.stopRecordingJob(state, callID); err != nil {
+				p.LogError("failed to stop recording job", "callID", callID, "err", err.Error())
+			}
+		}
+
+		p.publishWebSocketEvent(wsEventCallTranscriptionState, map[string]interface{}{
+			"callID":  callID,
+			"trState": clientState.toMap(),
+		}, &model.WebsocketBroadcast{ChannelId: callID, ReliableClusterSend: true})
+	}
+}
+
+func (p *Plugin) startTranscribingJob(state *channelState, callID, userID, trID string) (rerr error) {
+	if state.Call.Transcription != nil && state.Call.Transcription.EndAt == 0 {
+		return fmt.Errorf("transcription already in progress")
+	}
+
+	if trID == "" {
+		return fmt.Errorf("trID should not be empty")
+	}
+
+	trState := new(jobState)
+	trState.ID = trID
+	trState.CreatorID = userID
+	trState.InitAt = time.Now().UnixMilli()
+	state.Call.Transcription = trState
+
+	if err := p.kvSetChannelState(callID, state); err != nil {
+		return fmt.Errorf("failed to set channel state: %w", err)
+	}
+
+	defer func() {
+		// In case of any error we relay it to the client.
+		if rerr != nil && trState != nil {
+			trState.EndAt = time.Now().UnixMilli()
+			trState.Err = rerr.Error()
+			p.publishWebSocketEvent(wsEventCallTranscriptionState, map[string]interface{}{
+				"callID":  callID,
+				"trState": trState.getClientState().toMap(),
+			}, &model.WebsocketBroadcast{ChannelId: callID, ReliableClusterSend: true})
+		}
+	}()
+
+	// Sending the event prior to making the API call to the job service
+	// since it could take a few seconds to complete and we want clients
+	// to get their local state updated as soon as it changes on the server.
+	p.publishWebSocketEvent(wsEventCallTranscriptionState, map[string]interface{}{
+		"callID":  callID,
+		"trState": trState.getClientState().toMap(),
+	}, &model.WebsocketBroadcast{ChannelId: callID, ReliableClusterSend: true})
+
+	// We don't want to keep the lock while making the API call to the service since it
+	// could take a while to return. We lock again as soon as this returns.
+	p.unlockCall(callID)
+	trJobID, jobErr := p.getJobService().RunJob(job.TypeTranscribing, callID, state.Call.PostID, trState.ID, p.botSession.Token)
+	state, err := p.lockCall(callID)
+	if err != nil {
+		return fmt.Errorf("failed to lock call: %w", err)
+	}
+
+	trState, err = state.getTranscription()
+	if err != nil {
+		return fmt.Errorf("failed to get transcription state: %w", err)
+	}
+
+	if jobErr != nil {
+		state.Call.Transcription = nil
+		if err := p.kvSetChannelState(callID, state); err != nil {
+			return fmt.Errorf("failed to set channel state: %w", err)
+		}
+		return fmt.Errorf("failed to create transcription job: %w", jobErr)
+	}
+
+	if trState.JobID != "" {
+		return fmt.Errorf("transcription job already in progress")
+	}
```

Comment:
That's funny. Not sure how that would happen :)

Reply:
It actually can happen because we are not 100% atomic there due to the fact that we need to unlock to make the API call to the job service. So concurrent requests to start transcriptions could theoretically result in multiple jobs starting. Of course the client shouldn't ever do that but it's a possibility we may want to account for in case of issues/bugs.

Reply:
Ah, good to know, thanks